app/controllers/RequiredGiinController.scala

```diff
@@ -46,25 +47,34 @@ class RequiredGiinController @Inject() (
 
   val form: Form[String] = formProvider()
 
-  def onPageLoad(mode: Mode): Action[AnyContent] = (identify andThen getData) {
+  def onPageLoad(mode: Mode): Action[AnyContent] = (identify andThen getData andThen RequireData) {
     implicit request =>
-      val preparedForm = request.userAnswers.flatMap(_.get(RequiredGiinPage)) match {
+      val preparedForm = request.userAnswers.get(RequiredGiinPage) match {
         case None        => form
         case Some(value) => form.fill(value)
       }
-
-      Ok(view(preparedForm, mode, "fiName"))
+      request.userAnswers.get(ValidXMLPage) match {
+        case Some(validatedFileData) =>
+          Ok(view(preparedForm, mode, validatedFileData.messageSpecData.fiNameFromFim))
+        case _ => Redirect(controllers.routes.JourneyRecoveryController.onPageLoad())
+      }
   }
 
-  def onSubmit(mode: Mode): Action[AnyContent] = (identify andThen getData).async {
+  def onSubmit(mode: Mode): Action[AnyContent] = (identify andThen getData andThen RequireData).async {
     implicit request =>
       form
         .bindFromRequest()
         .fold(
-          formWithErrors => Future.successful(BadRequest(view(formWithErrors, mode, "fiName"))),
+          formWithErrors =>
+            request.userAnswers.get(ValidXMLPage) match {
```

Comment:
just curious, why do we need same validation as onpageload for submit as well?

Reply:
thanks got it but it would be better to get the validxmlpage and let system throw exception if its not available as it is technical error rather journey recovery